File: leetcode/Go/hashmaps/238/238.go
package main

func productExceptSelf(nums []int) []int {
	ans := make([]int, len(nums))
	zeroes := 0 // [index];
	counter := 1
	// Getting all numbers multiplied
	for _, num := range nums {
		if num == 0 {
			zeroes++
		} else {
			counter *= num
		}
	}

	for i, num := range nums {
		if num != 0 {
			// If at least one zero exists
			if zeroes >= 1 {
				ans[i] = 0
			} else {
				ans[i] = counter / num
			}
		} else {
			// If more than one zero exists
			if zeroes > 1 {
				ans[i] = 0
			} else {
				ans[i] = counter
			}
		}
	}
	return ans
}

File: leetcode/Python/arrays/238.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        solution = []
        totalValue = 1
        zeroes = 0
        for n in nums:
            if n != 0:
                totalValue *= n
            else: 
                zeroes += 1
        # In case all arrays are zero
        if zeroes > 1:
            return [0] * len(nums)

        # Managing cases with one zero
        for n in nums:
            if n != 0 and zeroes > 0:
                solution.append(0)
            elif n == 0 and zeroes > 0:
                solution.append(totalValue)
        if zeroes > 0:
            return solution
        for n in nums:
            solution.append(totalValue/n)
        return solution
        
